### ISSUE TITLE
Add function to wait till all sinks are finished

### ISSUE THREAD

**Issue by amarcum** [KjellKod/g3log#106]
When writing Google Tests it would be really nice to be able to call a function that will hault until all sinks are finished writing so that the log messages show up where the failure happened. This is really only a feature request because the logging is so much faster than any console output methods.

I know this is a weird request that would really only be used during Unit Testing but it would allow people that Unit Test the ability to still use the LOG functionality to log to a console to trace down a logical issue. If you feel this needs to be done in the sinks instead of globally then I can look into that as well.

**Comment by KjellKod** [comment]
This was addressed in the closed pull request: https://github.com/KjellKod/g3log/pull/105

I'm planning to re-open the pull request but without the unsafe danling reference that the closed pull request demanded.

**Comment by KjellKod** [comment]
this is on hold for now. I don't have a strong enough use case for adding this functionality. Please chime in if you feel that you do

**Comment by hoditohod** [comment]
Hi,
I'd like to have this feature for the same reason @amarcum wrote: writing unit tests with gtest and g3log will result a console output where gtest output and testcase output will be incorrectly interleaved. Having a g3log call in test teardowns that says "wait till sinks are complete" would be really nice. The test report tool I use associates test outputs with testcases, but right now it's useless.

BTW, I have a custom sink defined which prints to stdout and flushes the stream after each line, but this is not enough, as it's up to the scheduler to decide when to run the logwriter thread and the logs still don't interleave correctly. I also added an empty function to the sink, and placed a sync call via the SinkHandle (`sinkhandle->call(&MySink, nop)->wait();`). This effectively waits for processing of all previous logs in the sink's `Active` queue, and seems to work. But as I see logs go through another queue in `LogWorkerImpl`, and I can't synchronize on that. Was it really necessary to spawn 2 new threads just for logging?

Long story short, it would be nice to have this feature in the lib.

**Comment by KjellKod** [comment]
Each sink has its own queue and its own thread.   If they would share queue and thread it would give rise to another heap of troubles async/synchronization issues. 

The way you suggested to call the `sinkhandle` works. 

There was a PR a long time ago which got closed without merging. It had the functionality to remove sinks which would also flush their queues.  Would that be of interest (to re-open) to address this feature request?

**Comment by TomBecker-BD** [comment]
It is really important for a unit test to wait for all the threads it starts. Otherwise errors can occur in the test execution engine after it thinks the test is done. In my case I was also having a stupid problem where the test execution engine would never exit because I never shutdown logging. 

(I'm using Microsoft::VisualStudio::CppUnitTestFramework in Visual Studio 2015 but I think the issue is the same for all frameworks.) 

I tried creating a monitor class to init/shutdown logging for each test: 
```c++
std::unique_ptr<g3::LogWorker> worker;
std::unique_ptr<g3::SinkHandle<CoutSink>> consoleSink;

TempConsoleLog::TempConsoleLog(const LEVELS minLevel)
{
    worker = g3::LogWorker::createLogWorker();
    consoleSink = worker->addSink(std2::make_unique<CoutSink>(), &CoutSink::ReceiveLogMessage);
    g3::initializeLogging(worker.get());
    g3::log_levels::setHighest(minLevel);
}

TempConsoleLog::~TempConsoleLog()
{
    if (worker)
    {
        g3::internal::shutDownLoggingForActiveOnly(worker.get());
        worker.reset();
     }
    consoleSink.reset();
}
```
`~Active` joins the thread for the sink so I think all the async activity is finished deterministically when `worker` is destructed. I am not sure if any new API is needed. 

There is one problem: When I run tests under the debugger, the test execution engine has an unhandled access violation after all the tests have completed. Is there a better sequence I could use to shutdown logging so the test execution engine can exit cleanly?

**Comment by KjellKod** [comment]
I hear you guys.

There was a PR a long time ago which started off in the wrong direction,..
got on the right track after some input from me but the guy got pissed off
when I didn't agree with a design decision he made and he revoked some of
his code and closed the PR (*revoking does nothing of course since it's all
in git history*)

The PR is here: https://github.com/KjellKod/g3log/pull/105

If someone wants to take that up again I could quickly explain what needs
to be done to make it work. At this point I have no free  time to spend on
g3log except code reviews, testing and some guidence towards PR work.


* Contact Information
* Kjell Hedström
* Mobile:  +1 - 617.378.1915
* E-mail: Hedstrom@KjellKod.cc

**Comment by KjellKod** [comment]
There was a PR a long time ago which started off in the wrong direction,..got on the right track after some input from me but the guy got pissed off when I didn't agree with a design decision he made and he revoked some of his code and closed the PR ( _revoking does nothing of course since it's all in git history_ )

The PR is here: https://github.com/KjellKod/g3log/pull/105

If someone wants to take that up again I could quickly explain what needs to be done to make it work. At this point I have no free  time to spend on g3log except code reviews, testing and some guidence towards PR work.

**Comment by TomBecker-BD** [comment]
Maybe I am missing something but I don't think I want to remove the sink -- that would mean keeping the log worker around between tests. 

I like the approach I took with the monitor class because it did not require any API changes. I'm probably doing it wrong, because of the access violation, but in theory, it should be possible to shut down the log worker cleanly leaving the system in the state it was before logging was initialized.

**Comment by KjellKod** [comment]
No activity. Closing this.
